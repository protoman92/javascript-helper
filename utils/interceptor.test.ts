import createInterceptorRegistry from "./interceptor";

describe("Generic interceptor", () => {
  it("Add and remove interceptors should work", async () => {
    // Setup
    const registry = createInterceptorRegistry<(...args: any[]) => number>({
      resultCombiner: (...args) => args.reduce((acc, v = 0) => acc + v, 0),
    });

    const intercept1 = jest.fn();
    const intercept2 = jest.fn();
    intercept1.mockResolvedValueOnce(0);
    intercept2.mockResolvedValueOnce(0);

    // When
    registry.addInterceptor(intercept1);
    registry.addInterceptor(intercept2);
    const result1 = await registry.intercept(1, 2, 3);
    registry.removeInterceptor(intercept1);
    registry.removeInterceptor(intercept2);
    const result2 = await registry.intercept(1, 2, 3);

    // Then
    expect(registry.interceptorType).toBeInstanceOf(Function);
    expect(intercept1).toHaveBeenCalledTimes(1);
    expect(intercept2).toHaveBeenCalledTimes(1);
    expect(intercept1.mock.calls).toMatchSnapshot();
    expect(intercept2.mock.calls).toMatchSnapshot();
    expect(result1).toEqual(0);
    expect(result2).toEqual(undefined);
  });
});
