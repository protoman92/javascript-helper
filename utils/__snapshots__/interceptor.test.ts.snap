// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic interceptor Add and remove interceptors should work 1`] = `
Array [
  Array [
    Object {
      "originalArguments": Array [
        1,
        2,
        3,
      ],
      "previousResult": undefined,
    },
  ],
]
`;

exports[`Generic interceptor Add and remove interceptors should work 2`] = `
Array [
  Array [
    Object {
      "originalArguments": Array [
        1,
        2,
        3,
      ],
      "previousResult": 0,
    },
  ],
]
`;
